packages/modules/kostal_plenticore/device.py

```diff
@@ -0,0 +1,157 @@
+                        bat_state.power = - raw_inv_power + inverter_state.power - home_consumption
+                    else:
+                        # Wenn die Batterie geladen wird, dann ist PV-Leistung die Wechselrichter-AC-Leistung + die
+                        # Ladeleistung der Batterie. Die PV-Leistung ist die Summe aus verlustbehafteter
+                        # AC-Leistungsabgabe des WR und der DC-Ladeleistung. Die Wandlungsverluste werden also nur
+                        # in der PV-Leistung ersichtlich.
+                        inverter_state.power += bat_state.power
+            if set_inverter_state:
+                component.set(inverter_state)
+        elif isinstance(component, KostalPlenticoreCounter):
+            component.update(reader)
+    if bat_state:
+        for component in components:
+            if isinstance(component, KostalPlenticoreBat):
+                component.set(bat_state)
+    if set_inverter_state is False:
+        return inverter_state
+
+
+def create_device(device_config: KostalPlenticore):
+    def create_bat_component(component_config):
+        return KostalPlenticoreBat(device_config.id, component_config)
+
+    def create_counter_component(component_config):
+        return KostalPlenticoreCounter(device_config.id, component_config)
+
+    def create_inverter_component(component_config):
+        return KostalPlenticoreInverter(component_config)
+
+    def update_components(
+        components: Iterable[Union[KostalPlenticoreBat, KostalPlenticoreCounter, KostalPlenticoreInverter]]
+    ):
+        with tcp_client:
+            update(components, reader)
+
+    tcp_client = modbus.ModbusTcpClient_(device_config.configuration.ip_address, 1502)
+    reader = functools.partial(tcp_client.read_holding_registers, unit=71, wordorder=Endian.Little)
+    return ConfigurableDevice(
+        device_config=device_config,
+        component_factory=ComponentFactoryByType(
+            bat=create_bat_component, counter=create_counter_component, inverter=create_inverter_component),
+        component_updater=MultiComponentUpdater(update_components),
+    )
+
+
+def _create_reader(tcp_client: modbus.ModbusTcpClient_) -> Callable[[int, modbus.ModbusDataType], Any]:
+    def little_endian_wordorder_reader(register: int, data_type: modbus.ModbusDataType):
+        return tcp_client.read_holding_registers(
+            register, data_type, unit=71, wordorder=Endian.Little)
+    return little_endian_wordorder_reader
+
+
+def read_legacy_inverter(ip1: str, ip2: str, battery: int, ip3: str, position: int) -> InverterState:
+    # in IP3 kann ein aufeinanderfolgende Liste enthalten sein "192.168.0.1-3"
+    log.debug("Kostal Plenticore: WR1: {}, WR2: {}, Battery: {}, WR3: {}".format(ip1, ip2, battery, ip3))
+    additional_ips = [filter("none".__ne__, chain([ip2], ipparser(ip3)))]
+    client_1 = modbus.ModbusTcpClient_(ip1, 1502)
+    reader_1 = _create_reader(client_1)
```

Comment:
Die Namen `client_1` und `read_1` sagen mir nicht so zu. Du könntest sie `client_main` nennen oder so. Noch besser wäre aber das kommt in einen kleineren Scope.

Das ganze hat mich dazu gebracht generell zu überlegen ob es möglich wäre die Funktion völlig anders zu strukturieren. Hiermit bin ich rausgekommen:
```python
def read_legacy_inverter(ip1: str, ip2: str, battery: int, ip3: str) -> InverterState:
    # in IP3 kann ein aufeinanderfolgende Liste enthalten sein "192.168.0.1-3"
    log.debug("Kostal Plenticore: WR1: %s, WR2: %s, WR3: %s, Battery: %s", ip1, ip2, ip3, battery)
    inverter_component = KostalPlenticoreInverter(KostalPlenticoreInverterSetup(id=1))

    def get_hybrid_inverter_state(ip: str) -> InverterState:
        battery_component = KostalPlenticoreBat(1, KostalPlenticoreBatSetup())
        with modbus.ModbusTcpClient_(ip) as client:
            return request_components(
                [inverter_component, battery_component], _create_reader(client), set_inverter_state=False
            )

    def get_standard_inverter_state(ip: str) -> InverterState:
        with modbus.ModbusTcpClient_(ip) as client:
            return inverter_component.update(_create_reader(client))

    def inverter_state_sum(a: InverterState, b: InverterState) -> InverterState:
        return InverterState(exported=a.exported + b.exported, power=a.power + b.power)

    inverter_state = reduce(
        inverter_state_sum,
        map(get_standard_inverter_state, filter("none".__ne__, chain([ip2], ipparser(ip3)))),
        get_hybrid_inverter_state(ip1) if battery else get_standard_inverter_state(ip1)
    )
    inverter_component.set(inverter_state)
    return inverter_state
```
Das sind 26 Zeilen statt 20 Zeilen vorher, also leider länger. Ist es das wert? Ist es besser lesbar? Das liegt wohl im Auge des Betrachters. Du kannst ja mal mit etwas Abstand drauf schauen ob du es verständlich findest.